File: insertionSortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if head == None or head.next == None:
            return head
            
        tmp = ListNode(0)
        tmp.next = head
        curr = head
        while curr.next:
            # Need change
            if curr.next.val < curr.val:
                pre = tmp
                while pre.next.val < curr.next.val:
                    pre = pre.next
                succ = curr.next
                curr.next = succ.next
                succ.next = pre.next
                pre.next = succ
            # normal, ascending
            else:
                curr = curr.next
        return tmp.next
            

File: CPP/Second/insertion_sort_list.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
        if (!head || !head->next)
            return head;
        ListNode *new_head = new ListNode(-1);
        new_head->next = head;
        ListNode *p = head;
        while (p->next) {
            if (p->next->val < p->val) {
                auto prev = new_head;
                while (prev->next->val <= p->next->val) 
                    prev = prev->next;
                // swap, succ is the one need to swap
                auto succ = p->next;
                p->next = succ->next;
                succ->next = prev->next;
                prev->next= succ;
            }
            else {
                p = p->next;
            }
        }
        return new_head->next;
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
